Buyback and distribute

SentinelDev1 opened the pull request.

SentinelDev1 committed:
Initial commit

-Updated FeeHandler.sol to allow a user to claim fees in exchange for GMX or WNT and added the ability to also claim V1 fees

-Added buyback related base keys and corresponding key getter functions to Keys.sol

-Added new errors used in FeeHandler.sol to Errors.sol

-Added IFeedAddress.sol interface used for calculating oracle prices

SentinelDev1 committed:
Correction to Keys.sol

-Adding back accidentally omitted base keys and getter functions related to the "contributor" updates

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -140,6 +140,8 @@ library Errors {
 
     // FeeHandler errors
     error InvalidClaimFeesInput(uint256 marketsLength, uint256 tokensLength);
+    error InvalidSwapTokenInput(address swapToken, uint256 arrayPosition);
```
can we call the token "buybackToken" and call the error "InvalidBuybackTokenInput"? i think that would be clearer to differentiate it from the regular GMX V2 swaps

SentinelDev1 replied:
I've updated accordingly in the latest commit

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,229 @@
     function claimFees(
-        address[] memory markets,
-        address[] memory tokens
+        address[] calldata markets,
```
i think we should change this to claimFees for only one market and one buybackToken, the logic seems a bit complicated for an array now

SentinelDev1 replied:
Updated to remove the loop in the latest commit

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,229 @@
+        address receiver = address(this);
+
+        (uint256 gmxTokenOraclePrice, uint256 gmxTokenOracleDecimals, uint256 gmxTokenDecimals) = _getTokenPricingValues(gmx);
+        (uint256 wntTokenOraclePrice, uint256 wntTokenOracleDecimals, uint256 wntTokenDecimals) = _getTokenPricingValues(wnt);
```
do we need to specify "gmx" and "wnt"? i had the impression that the user would just specify the buybackToken, and all the values would be fetched based on the buybackToken, without explicitly stating the token is gmx or wnt, if there is a buyback batchsize and buyback factor defined for that token, that would mean the token can be bought back, is that possible?

SentinelDev1 replied:
Updated so that gmx and wnt are only referenced in the _setAvailableFeeAmounts() private function. 

I'm not sure that it can be avoided in this case given that the buyback factor is defined for gmx and for wnt it is inferred. 

Also, in the claimFees() function the buyback token is no longer specified now that buybackFees() is it's own function, thus when computing the updated available fee amounts for gmx and wnt the tokens need to be explicitly referenced.

SentinelDev1 committed:
Updated FeeHandler.sol, Errors.sol and Keys.sol

-Updated FeeHandler.sol to inherit BasicMulticall.sol, moved the buyback of fees into a separate function named buybackFees() and added a withdrawFees() function to allow the feeKeeper to retreive fees from the FeeHandler contract

-Refactored FeeHandler.sol private functions and removed a few that were only used once after recent changes

-Updated FeeHandler related errors in Errors.sol and removed InvalidClaimFeesInput error now that arrays are not used in the FeeHandler claimFees() function

-Updated Keys.sol BUYBACK_MAX_SWAP_PRICE_IMPACT key name to BUYBACK_MAX_PRICE_IMPACT_FACTOR and the getter function name from buybackMaxSwapPriceImpactKey() to buybackMaxPriceImpactFactorKey()

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+    // @dev withdraw fees from this contract
+    // @param token the token for which to claim fees
+    // @param amount the amount of fees to claim
+    function withdrawFees(address token, uint256 amount) external nonReentrant onlyFeeKeeper {
```
would be good to validate that there is a buybackBatchAmount for the "token", to prevent accidental withdrawal of fee tokens that are not GMX and not WNT

SentinelDev1 replied:
No problem, I’ve added that validation. I had been thinking the same, but wasn't sure if we wanted the ability to withdraw tokens that weren't GMX or WNT in case of an emergency or for some other reason.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+        if (market == address(0)) {
+            version = 1;
+            buybackGmxFactor = _getBuybackGmxFactor(version);
+            feeAmount = IVaultV1(vaultV1).withdrawFees(feeToken, address(this));
```
for V1, only Vault.gov can withdraw fees, Vault.gov is a Timelock which also has a function withdrawFees, so this should be `IVaultGovV1(vaultV1.gov()).withdrawFees(feeToken, address(this))` instead

SentinelDev1 replied:
I've updated for this but I'm still assigning vaultV1 as an address in storage instead of as an contract/interface as is implied in the above example because vaultV1 needs to be passed as an address in the IVaultGovV1().withdrawFees() function.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+    function buybackFees(address feeToken, address buybackToken) external nonReentrant {
+        uint256 batchSize = dataStore.getUint(Keys.buybackBatchAmountKey(buybackToken));
+        if (batchSize == 0) {
+            revert Errors.InvalidBuybackTokenInput(buybackToken);
```
this error can be "InvalidBuybackToken" instead of "InvalidBuybackTokenInput"

SentinelDev1 replied:
No problem, I've updated accordingly.

SentinelDev1 replied:
I don't want to jump the gun, but I think this can be marked resolved now?

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+        address buybackToken,
+        uint256 batchSize
+    ) private view returns (uint256) {
+        (
```
should use https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/oracle/ChainlinkPriceFeedUtils.sol#L17 instead

the returned price would already be adjusted based on the token decimals

SentinelDev1 replied:
Thanks, I've updated to use this instead.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+        address tokenPriceFeedAddress = dataStore.getAddress(Keys.priceFeedKey(token));
+        uint256 tokenOraclePrice = IFeedAddress(tokenPriceFeedAddress).latestAnswer();
+        uint256 tokenOracleDecimals = IFeedAddress(tokenPriceFeedAddress).decimals();
+        uint256 tokenDecimals = IFeedAddress(token).decimals(); // IERC20().decimals() is not included in the openzeppelin IERC20 interface so using IFeedAddress().decimals() instead but can change if necessary
```
there should not be a need to read token decimals if the prices are already adjusted: https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/oracle/ChainlinkPriceFeedUtils.sol#L62

SentinelDev1 replied:
As per the above comment, this is no longer needed and has been removed.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+            revert Errors.InvalidBuybackTokenInput(buybackToken);
+        }
+
+        uint256 minFeeTokenAmount = _getMinFeeTokenAmount(feeToken, buybackToken, batchSize);
```
could call this expectedFeeTokenAmount

SentinelDev1 replied:
Updated to expectedFeeTokenAmount.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+        } else if (availableFeeAmount >= minFeeTokenAmount) {
+            _buybackFees(feeToken, buybackToken, batchSize, availableFeeAmount, availableFeeAmount);
+        } else {
+            revert Errors.InsufficientFeeAmount(feeToken, buybackToken, availableFeeAmount);
```
i don't think this is necessary, maybe let the user pass in a "minOutputAmount" and validate that the amount that would be sent to the user is >= minOutputAmount, error should be InsufficientBuybackOutputAmount to separate it from the regular fee and swap errors

SentinelDev1 replied:
Updated to add minOutputAmount as an input param and updated the error to InsufficientBuybackOutputAmount. The error InsufficientBuybackOutputAmount doesn’t include minOutputAmount as a param because it is passed by the user to the function so they should already have that information, but it can be added as well if preferred.

xvi10 replied:
would be good to add minOutputAmount to the error for convenience, in the Errors we usually always return (..., "the actual value", "the expected value")

SentinelDev1 replied:
Ok no problem, this has been added.

xvi10 commented on contracts/fee/IFeedAddress.sol:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: BUSL-1.1
```
this should not be needed, the Chainlink logic is already in https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/oracle/ChainlinkPriceFeedUtils.sol#L62

SentinelDev1 replied:
As per the above comment, this is no longer needed and has been removed.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+    // @dev receive an amount in feeToken by depositing the batchSize amount of the buybackToken
+    // @param feeToken the token to receive with the fee amount calculated via an oracle price
+    // @param buybackToken the token to deposit in the amount of batchSize in return for fees
+    function buybackFees(address feeToken, address buybackToken) external nonReentrant {
```
let's call this buyback instead of buybackFees

SentinelDev1 replied:
This has now been updated to buyback() from buybackFees().

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+    // @param feeToken the token to receive with the fee amount calculated via an oracle price
+    // @param buybackToken the token to deposit in the amount of batchSize in return for fees
+    function buybackFees(address feeToken, address buybackToken) external nonReentrant {
+        uint256 batchSize = dataStore.getUint(Keys.buybackBatchAmountKey(buybackToken));
```
we should also validate that feeToken and buybackToken are not the same, error can be BuybackAndFeeTokenAreEqual

SentinelDev1 replied:
This validation has now been added.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+    // @param market the markets to claim fees from
+    // @param feeToken the fee tokens to claim
+    function claimFees(address market, address feeToken) external nonReentrant {
+        uint256 version;
```
should validate that there is no buybackBatchAmount for the feeToken, GMX and WNT fees should not go through this claim and buyback process

SentinelDev1 replied:
This validation has now been added.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+    // @param token the token for which to claim fees
+    // @param amount the amount of fees to claim
+    function withdrawFees(address token, uint256 amount) external nonReentrant onlyFeeKeeper {
+        address receiver = dataStore.getAddress(Keys.FEE_RECEIVER);
```
should allow "markets[]" to be passed in and call FeeUtils.claimFees(dataStore, eventEmitter, market, token, address(this)), to allow all WNT or GMX to be directly claimed from the markets

SentinelDev1 replied:
withdrawFees has now been updated to include an address[] calldata markets input param. Currently, the buybackToken fees are first consolidated in FeeHandler and then transferred to FEE_RECEIVER. If preferred, the claimed fees could be transferred directly to FEE_RECEIVER from the GM market. 

Also, this could be updated to also include an array of buybackTokens to allow for both GMX and WNT to be withdrawn in one call, though I think this would require either matching the length of markets and buybackTokens or a nested for loop.

xvi10 replied:
the current logic looks ok to me

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,176 @@
+    // @param amount the amount of fees to claim
+    function withdrawFees(address token, uint256 amount) external nonReentrant onlyFeeKeeper {
+        address receiver = dataStore.getAddress(Keys.FEE_RECEIVER);
+        IERC20(token).safeTransfer(receiver, amount);
```
could change this to `IERC20(token).safeTransfer(receiver, IERC20(token).balanceOf(address(this));` so that amount does not need to be passed in

SentinelDev1 replied:
This has been updated accordingly. I had been thinking of taking this approach as well, but included the amount input param in case for some reason it would be useful to specify it explicitly.

SentinelDev1 committed:
Updated FeeHandler.sol, Errors.sol, IVaultV1.sol, created IVaultGovV1.sol and deleted IFeedAddress.sol

-Updates to FeeHandler.sol per the comments in the pull request and telegram group

-Updated existing FeeHandler errors and added new errors to Errors.sol

-Added gov() view function to IVaultV1.sol

-Added IVaultGovV1.sol interface with the withdrawFees() function

-Deleted IFeedAddress.sol interface as it is not needed

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,208 @@
+        _validateBuybackToken(batchSize, buybackToken);
+
+        uint256 availableFeeAmount = _getAvailableFeeAmount(feeToken, buybackToken);
+        uint256 maxFeeTokenAmount = _getMaxFeeTokenAmount(feeToken, buybackToken, batchSize);
```
i think this would be clearer as:
```
uint256 outputAmount = availableFeeAmount < maxFeeTokenAmount ? availableFeeAmount : maxFeeTokenAmount

if (buybackAmount < minOutputAmount) { revert ... }

_buybackFees(feeToken, buybackToken, batchSize, outputAmount, availableFeeAmount);
```

SentinelDev1 replied:
That makes sense, I've updated per the above suggestion.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,208 @@
+        IERC20(buybackToken).safeTransferFrom(msg.sender, address(this), batchSize);
+
+        availableFeeAmount = availableFeeAmount - buybackAmount;
+        _setAvailableFeeAmount(feeToken, buybackToken, availableFeeAmount);
```
may be slightly clearer as:

 _setAvailableFeeAmount(feeToken, buybackToken, availableFeeAmount - buybackAmount);

SentinelDev1 replied:
I agree, I’ve updated per your suggestion above.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,208 @@
+        _setAvailableFeeAmount(feeToken, buybackToken, availableFeeAmount);
+    }
+
+    function _setAvailableFeeAmounts(uint256 version, address feeToken, uint256 feeAmount) private {
```
would be clearer to name this _incrementAvailableFeeAmounts to differentiate it from _setAvailableFeeAmount

SentinelDev1 replied:
No problem, I’ve renamed the function for more clarity.

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,208 @@
+        uint256 batchSize = dataStore.getUint(Keys.buybackBatchAmountKey(buybackToken));
+        _validateBuybackToken(batchSize, buybackToken);
+
+        uint256 availableFeeAmount = _getAvailableFeeAmount(feeToken, buybackToken);
```
may be good to revert if availableFeeAmount is zero

SentinelDev1 replied:
Ok, no problem. I added another AvailableFeeAmountIsZero error that returns (address feeToken, address buybackToken, uint256 availableFeeAmount) since it didn't seem to make sense to return minOutputAmount as is now done in the InsufficientBuybackOutputAmount error.

SentinelDev1 committed:
Updated FeeHandler.sol, Config.sol and Errors.sol

-Refactored FeeHandler.sol to remove local variable declarations in cases where the variable was only referenced once and added a 2 more private functions for code reused multiple times.

-Now using the Precision.sol library in FeeHandler.sol for FLOAT_PRECISION as well as the mulDiv() and applyFactor() functions where applicable

-Added BUYBACK_BATCH_AMOUNT, BUYBACK_GMX_FACTOR and BUYBACK_MAX_PRICE_IMPACT_FACTOR to the _initAllowedBaseKeys() function in Config.sol

-Added BUYBACK_MAX_PRICE_IMPACT_FACTOR and BUYBACK_GMX_FACTOR to the _validateRange() function in Config.sol

-Added the AvailableFeeAmountIsZero() error to Errors.sol

SentinelDev1 committed:
Updated Config.sol

-Changed the range validation for Keys.BUYBACK_MAX_PRICE_IMPACT_FACTOR from 'revert if value > 1%' to 'revert if value > 5%'

gdev8317 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,215 @@
+        address buybackToken
+    ) external view returns (uint256) {
+        uint256 batchSize = _getBatchSize(buybackToken);
+        _validateBuybackToken(batchSize, buybackToken);
```
small suggestion, i think we usually don't use prefix "_" for private/internal methods, probably makes sense not to use it here for consistency?

xvi10 replied:
we do use prefix "_" for private / internal methods: https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/exchange/OrderHandler.sol#L352

i think only in libraries do we not do that

gdev8317 replied:
i see, i had an impression that this prefix is an exception and rarely used

gdev8317 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,215 @@
+    }
+
+    function _setAvailableFeeAmount(address feeToken, address buybackToken, uint256 availableFeeAmount) private {
+        dataStore.setUint(Keys.buybackAvailableFeeAmountKey(feeToken, buybackToken), availableFeeAmount);
```
also wondering if it's necessary to create helpers that only call setUint/getUint, at first glance calling dataStore directly is clear/concise as well?

xvi10 replied:
i think this is fine, we do it in multiple other places as well: https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/market/MarketUtils.sol#L1986

gdev8317 replied:
maybe it's fine if these methods are later used in some other contracts, but in this case it only used inside this contract

it might be a matter of taste, i would personally prefer not to add more helpers, in this specific case they don't make code clearer in my view, but increase the code source size / adds some complexity to understand that code because i need to do an extra step to check what's inside `_setAvailableFeeAmount`

xvi10 replied:
discussed that this can be separately decided on

gdev8317 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,215 @@
+    // @param market the market from which to claim fees
+    // @param feeToken the fee tokens to claim from the market
+    function claimFees(address market, address feeToken) external nonReentrant {
+        if (_getBatchSize(feeToken) != 0) {
```
just checking should it be `== 0` instead?

gdev8317 replied:
like here https://github.com/gmx-io/gmx-synthetics/pull/53/files#diff-537965269a61a84567326928bad5ef66fa8b7a71cff643c237907d9dba0e7e17R213

xvi10 replied:
"!= 0" is correct, we do not want the user to be able to claim fees for the buyback tokens (GMX, WNT)

could add a comment to make it clearer

gdev8317 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,215 @@
+        }
+
+        uint256 feeAmount;
+        if (market == address(0)) {
```
i would suggest for `claimFees` func to accept `uint version` arg to make it more explicit and not rely on market

gdev8317 replied:
and maybe create constants like `const uint V1 = 1` and use them instead of plain values 1 and 2

xvi10 replied:
ok with me

SentinelDev1 replied:
Ok, I've updated per G's recommendation above in my latest commit, including adding v1 and v2 const uint256 variables

gdev8317 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,215 @@
+        uint256 feeTokenPrice = _getPriceFeedPrice(feeToken);
+        uint256 buybackTokenPrice = _getPriceFeedPrice(buybackToken);
+
+        uint256 expectedFeeTokenAmount = Precision.mulDiv(batchSize, feeTokenPrice, buybackTokenPrice);
```
do i understand correctly, that buybackToken is GMX? in this case i guess it should be
```
Precision.mulDiv(batchSize, buybackTokenPrice, feeTokenPrice);
```

otherwise if batch size 100, GMX token price is $20 and fee token = USDC then expected fee token amount = 100 * 1 / 20 = 5 USDC. and i guess it should be 2000 USDC?

xvi10 replied:
i think you're right, this should be changed

SentinelDev1 replied:
Yup, this was also identified in the audit as one of the findings and has been corrected in the latest commit

SentinelDev1 committed:
Updated FeeHandler.sol, Keys.sol and Errors.sol

- Updated FeeHandler.sol to use prices from Oracle.sol and validate that the price is not older than BUYBACK_MAX_PRICE_AGE, updated maxPriceImpactFactor to also include the maxPriceImpactFactor for the buybackToken in _getMaxFeeTokenAmount(), corrected expectedFeeTokenAmount calculation in _getMaxFeeTokenAmount(), moved safeTransferFrom() to occur before safeTransfer() in _buybackFees()

- Added uint256 constants v1 and v2, uint256 version input param to claimFees() and getOutputAmount() and Oracle _oracle input param to the constructor in FeeHandler.sol

- Added BUYBACK_MAX_PRICE_AGE key and buybackMaxPriceAgeKey() getter function to Keys.sol

- Added errors InvalidVersion() and PricesOlderThanMaxPriceAge() to Errors.sol

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -139,7 +139,13 @@ library Errors {
+    error InsufficientBuybackOutputAmount(address feeToken, address buybackToken, uint256 outputAmount, uint256 minOutputAmount);
+    error BuybackAndFeeTokenAreEqual(address feeToken, address buybackToken);
+    error AvailableFeeAmountIsZero(address feeToken, address buybackToken, uint256 availableFeeAmount);
+    error PricesOlderThanMaxPriceAge(uint256 priceTimestamp, uint256 buybackMaxPriceAge, uint256 currentTimestamp);
```
Could be named MaxBuybackPriceAgeExceeded

SentinelDev1 replied:
Ok, this has been renamed in the latest commit

xvi10 commented on contracts/data/Keys.sol:
```diff
@@ -1998,4 +2000,10 @@ library Keys {
+
+    // @dev key for the buyback max price age
+    // @return key for buyback max price age
+    function buybackMaxPriceAgeKey() internal pure returns (bytes32) {
```
if there is no params required for the key, then usually we just refer to it directly, so a function is not needed, e.g.: https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/oracle/Oracle.sol#L244

SentinelDev1 replied:
Ok, this has been removed in the latest commit and the base key is now referenced directly in FeeHandler.sol

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -60,26 +61,33 @@ contract FeeHandler is ReentrancyGuard, RoleModule, BasicMulticall {
-            _incrementAvailableFeeAmounts(1, feeToken, feeAmount);
-        } else {
+        } else if (version == v2) {
             feeAmount = FeeUtils.claimFees(dataStore, eventEmitter, market, feeToken, address(this));
```
should validate that market is not address(0) here

SentinelDev1 replied:
Ok, this has been added in the latest commit

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,221 @@
+    }
+
+    function _setAvailableFeeAmount(address feeToken, address buybackToken, uint256 availableFeeAmount) private {
+        dataStore.setUint(Keys.buybackAvailableFeeAmountKey(feeToken, buybackToken), availableFeeAmount);
```
should emit a SetAvailableFeeAmount event here

SentinelDev1 replied:
Ok, this has been added in the latest commit, not sure if buybackToken (and maybe feeToken) should be indexed?

xvi10 commented on contracts/fee/FeeHandler.sol:
```diff
@@ -3,46 +3,221 @@
+            revert Errors.InsufficientBuybackOutputAmount(feeToken, buybackToken, outputAmount, minOutputAmount);
+        }
 
+        _buybackFees(feeToken, buybackToken, batchSize, outputAmount, availableFeeAmount);
```
should emit a Buyback event here

SentinelDev1 replied:
Ok, this has been added in the latest commit, not sure if buybackToken (and maybe feeToken) should be indexed?

xvi10 replied:
i think it is fine without index

xdev10 committed:
add withdrawable buyback token amount

xdev10 committed:
add FeeHandler.getOutputAmount

SentinelDev1 committed:
Updated FeeHandler.sol, Keys.sol and Errors.sol

- Updated FeeHandler.sol to emit events for SetAvailableFeeAmount and BuybackFees and now validating whether market = address(0) in claimFees() and getOutputAmount() if version = v2

- Removed the unnecessary buybackMaxPriceAgeKey() function from Keys.sol and now just directly referencing the BUYBACK_MAX_PRICE_AGE base key in FeeHandler.sol

- Renamed error PricesOlderThanMaxPriceAge to MaxBuybackPriceAgeExceeded and added error EmptyClaimFeesMarket

xdev10 committed:
Merge branch 'updates' into buyback-and-distribute

# Conflicts:
#	contracts/config/Config.sol

xdev10 committed:
add note about FeeHandler.getOutputAmount

SentinelDev1 committed:
Updated deployFeeHandler.ts, added mock contracts and mock contract deployment scripts

SentinelDev1 committed:
Added FeeHandler.ts test script and other related updates

xdev10 committed:
optimize fee amounts

SentinelDev1 committed:
Updated FeeHandler.ts script and a few other related files

- Updated the FeeHandler test to validate V1 fees for all functions, added claiming fees in a buyback token to validate WithdrawableBuybackTokenAmount logic, and added a check for availableFeeAmount at the end of the test to validate those values

- Removed testFees FeeHandler test since it is no longer applicable

- Corrected logic in MockVaultV1 feeReserves() function

xdev10 committed:
Merge branch 'updates' into buyback-and-distribute

# Conflicts:
#	config/oracle.ts
